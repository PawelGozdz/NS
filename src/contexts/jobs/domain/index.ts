export * from './job-position';
export * from './job-user-profile';
export * from './jobs';
export * from './skill';
