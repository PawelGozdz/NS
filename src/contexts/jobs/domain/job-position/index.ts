export * from './errors';
export * from './events';
export * from './job-position-command-repository.interface';
export * from './job-position-entity.fixture';
export * from './job-position-query-repository.interface';
export * from './job-position.entity';
export * from './job-position.snapshot';
