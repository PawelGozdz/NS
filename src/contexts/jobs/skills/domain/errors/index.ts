export * from './skill-already-exists.error';
export * from './skill-not-found.error';
