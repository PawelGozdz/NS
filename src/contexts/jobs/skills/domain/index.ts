export * from './errors';
export * from './events';
export * from './skill-command-repository.interface';
export * from './skill-entity.fixture';
export * from './skill-query-repository.interface';
export * from './skill.entity';
