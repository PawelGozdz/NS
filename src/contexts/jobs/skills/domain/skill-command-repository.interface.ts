import { ISkillCreateData, Skill } from './skill.entity';

export abstract class ISkillsCommandRepository {
  abstract save(skill: ISkillCreateData): Promise<{ id: number }>;

  abstract getOneById(id: number): Promise<Skill | undefined>;

  abstract getOneByNameAndCategoryId(name: string, categoryId: number): Promise<Skill | undefined>;
}
