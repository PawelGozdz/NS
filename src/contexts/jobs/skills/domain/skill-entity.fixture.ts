/* eslint-disable @typescript-eslint/strict-boolean-expressions */
import { Skill } from './skill.entity';

type Input = {
  id?: number;
  name?: string;
  description?: string;
  categoryId?: number;
};

export class SkillEntityFixtureFactory {
  public static create(overrides?: Input): Skill {
    const id = overrides?.id ?? 1;
    const name = overrides?.name ? overrides.name : 'Grocery';
    const description = overrides?.description ? overrides.description : 'Grocery Skill';
    const categoryId = overrides?.categoryId ? overrides.categoryId : 3;

    return new Skill({
      id,
      name,
      description,
      categoryId,
    });
  }
}
