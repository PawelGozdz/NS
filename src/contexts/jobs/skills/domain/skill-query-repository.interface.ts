import { ISkillsQueryParams } from '@app/core';

export type SkillInfo = {
  id: number;
  name: string;
  context: string;
  parentId: number | null;
  description: string | null;
  categoryId: number;
};

export abstract class ISkillsQueryRepository {
  abstract getManyBy(query: ISkillsQueryParams): Promise<SkillInfo[]>;
}
