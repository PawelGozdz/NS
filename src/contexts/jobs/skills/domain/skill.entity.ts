export class Skill {
  id: number;

  name: string;

  description?: string | null;

  parentId?: number | null;

  context: string;

  categoryId: number;

  constructor(props: Skill) {
    this.id = props.id;
    this.name = props.name;
    this.description = props.description ?? null;
    this.parentId = props.parentId ?? null;
    this.context = props.context;
    this.categoryId = props.categoryId;
  }

  public static create({
    id,
    name,
    description,
    parentId,
    context,
    categoryId,
  }: {
    id: number;
    name: string;
    description?: string | null;
    parentId?: number | null;
    context: string;
    categoryId: number;
  }): Skill {
    const skill = new Skill({
      id,
      name,
      description: description ?? null,
      parentId: parentId ?? null,
      context,
      categoryId,
    });

    return skill;
  }
}

export type ISkillCreateData = {
  name: string;
  description?: string | null;
  parentId?: number | null;
  context: string;
  categoryId: number;
};

export type ISkillUpdateData = {
  id: number;
  name?: string;
  description?: string | null;
  context?: string;
  parentId?: number | null;
  categoryId?: number;
};
