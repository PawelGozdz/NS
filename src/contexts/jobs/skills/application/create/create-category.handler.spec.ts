import { createMock } from '@golevelup/ts-jest';
import { Test } from '@nestjs/testing';

import { IOutboxRepository } from '@app/core';
import { QueryBus } from '@libs/cqrs';
import { TestCqrsModule, TestLoggerModule, catchActError } from '@libs/testing';

import { ISkillsCommandRepository, SkillAlreadyExistsError, SkillEntityFixtureFactory, SkillNotFoundError } from '../../domain';
import { CreateSkillCommand } from './create-skill.command';
import { CreateSkillHandler } from './create-skill.handler';

jest.mock('@nestjs-cls/transactional', () => ({
  Transactional: () => jest.fn(),
}));

describe('CreateSkillHandler', () => {
  let handler: CreateSkillHandler;
  let skillCommandRepositoryMock: jest.Mocked<ISkillsCommandRepository>;
  let outboxRepositoryMock: jest.Mocked<IOutboxRepository>;
  let queryBusMock: jest.Mocked<QueryBus>;

  beforeEach(async () => {
    skillCommandRepositoryMock = createMock();
    outboxRepositoryMock = createMock();
    queryBusMock = createMock();

    const app = await Test.createTestingModule({
      imports: [TestLoggerModule.forRoot(), TestCqrsModule],
      providers: [
        CreateSkillHandler,
        {
          provide: ISkillsCommandRepository,
          useValue: skillCommandRepositoryMock,
        },
        {
          provide: IOutboxRepository,
          useValue: outboxRepositoryMock,
        },
        {
          provide: QueryBus,
          useValue: queryBusMock,
        },
      ],
    }).compile();

    handler = app.get(CreateSkillHandler);
  });

  const categoryId = 5;
  const name = 'test';
  const id = 1;
  const command = new CreateSkillCommand({
    name,
    categoryId,
  });

  const category = { id: categoryId };

  const entity = SkillEntityFixtureFactory.create();

  describe('Success', () => {
    it('should create new skill', async () => {
      // Arrange
      skillCommandRepositoryMock.getOneByNameAndCategoryId.mockResolvedValueOnce(undefined);
      queryBusMock.execute.mockResolvedValueOnce(category);
      skillCommandRepositoryMock.save.mockResolvedValueOnce({ id });

      // Act
      const result = await handler.execute(command);

      // Assert

      expect(skillCommandRepositoryMock.save).toHaveBeenCalledWith({
        name,
        description: undefined,
        categoryId,
      });
      expect(outboxRepositoryMock.store).toHaveBeenCalledTimes(1);
      expect(result).toMatchSnapshot();
    });
  });

  describe('failure', () => {
    it('should throw SkillAlreadyExistsError', async () => {
      // Arrange
      skillCommandRepositoryMock.getOneByNameAndCategoryId.mockResolvedValueOnce(entity);

      // Act
      const { error } = await catchActError(() => handler.execute(command));

      // Assert

      expect(skillCommandRepositoryMock.save).toHaveBeenCalledTimes(0);
      expect(outboxRepositoryMock.store).toHaveBeenCalledTimes(0);
      expect(error).toBeInstanceOf(SkillAlreadyExistsError);
      expect(error).toMatchSnapshot();
    });

    it('should throw SkillNotFoundError', async () => {
      // Arrange
      skillCommandRepositoryMock.getOneByNameAndCategoryId.mockResolvedValueOnce(undefined);
      queryBusMock.execute.mockResolvedValueOnce([]);

      // Act
      const { error } = await catchActError(() => handler.execute(command));

      // Assert

      expect(skillCommandRepositoryMock.save).toHaveBeenCalledTimes(0);
      expect(outboxRepositoryMock.store).toHaveBeenCalledTimes(0);
      expect(error).toBeInstanceOf(SkillNotFoundError);
      expect(error).toMatchSnapshot();
    });
  });
});
