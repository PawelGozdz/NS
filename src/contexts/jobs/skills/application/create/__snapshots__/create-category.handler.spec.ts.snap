// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateSkillHandler Success should create new skill 1`] = `
{
  "id": 1,
}
`;

exports[`CreateSkillHandler failure should throw SkillAlreadyExistsError 1`] = `[SkillAlreadyExistsError: Entity with name test and categoryId 5 already exists]`;

exports[`CreateSkillHandler failure should throw SkillNotFoundError 1`] = `[SkillNotFoundError: Category with id 5 doesn't exists]`;
